include/wil/safearrays.h

```diff
@@ -0,0 +1,1105 @@
+        {
+            if (psa != nullptr)
+            {
+                ULONGLONG elements = 1;
+                for (UINT i = 0; i < psa->cDims; ++i)
+                {
+                    // Determine the total size of the safearray by multiplying
+                    // the sizes of each dimension together
+                    elements *= psa->rgsabound[i].cElements;
+                    if (elements > ULONG_MAX)
+                    {
+                        RETURN_HR(HRESULT_FROM_WIN32(ERROR_ARITHMETIC_OVERFLOW));
+                    }
+                }
+                *result = static_cast<ULONG>(elements);
+            }
+            else
+            {
+                // If it's invalid, it doesn't contain any elements
+                *result = 0;
+            }
+            return S_OK;
+        }
+
+        typedef resource_policy<SAFEARRAY*, decltype(&details::SafeArrayDestroy), details::SafeArrayDestroy, details::pointer_access_all> safearray_resource_policy;
+    }
+    /// @endcond
+
+    typedef unique_any<SAFEARRAY*, decltype(&details::SafeArrayUnlock), details::SafeArrayUnlock, details::pointer_access_noaddress> safearray_unlock_scope_exit;
+    typedef unique_any<SAFEARRAY*, decltype(&details::SafeArrayUnaccessData), details::SafeArrayUnaccessData, details::pointer_access_noaddress> safearray_unaccess_scope_exit;
+
+    //! Guarantees a SafeArrayUnlock call on the given object when the returned object goes out of scope
+    //! Note: call SafeArrayUnlock early with the reset() method on the returned object or abort the call with the release() method
+    WI_NODISCARD inline safearray_unlock_scope_exit SafeArrayUnlock_scope_exit(SAFEARRAY* psa) WI_NOEXCEPT
+    {
+        details::SafeArrayLock(psa);
+        return safearray_unlock_scope_exit(psa);
+    }
+
+    //! Guarantees a SafeArrayUnaccessData call on the given object when the returned object goes out of scope
+    //! Note: call SafeArrayUnaccessData early with the reset() method on the returned object or abort the call with the release() method
+    template<typename T>
+    WI_NODISCARD inline safearray_unaccess_scope_exit SafeArrayAccessData_scope_exit(SAFEARRAY* psa, T*& p) WI_NOEXCEPT
+    {
+        details::SafeArrayAccessData(psa, p);
+        return safearray_unaccess_scope_exit(psa);
+    }
+
+    //! Class that facilitates the direct access to the contents of a SAFEARRAY and "unaccessing" it when done
+    //! It enables treating the safearray like an regular array and/or performing a ranged-for on the contents.
+    //! Accessing a safearray increases it's lock count, so attempts to destroy the safearray will fail until it is unaccessed.
+    //! This class works even if the SAFEARRAY is multi-dimensional by treating it as a large single-dimension array,
+    //! but doesn't support access via a multi-dimensional index.
+    //! NOTE: This class does not manage the lifetime of the SAFEARRAY itself.  See @ref safearray_t.

```

Comment:
If it's non-owning, the name should indicate that better. In C++, the names `view` (non-modifiable) or `span` (modifiable) are typically used to indicate that

Reply:
It is owning.  When SafeArrayAccessData is invoked, it creates a refcount and a pointer that this class owns until SafeArrayUnaccessData is invoked.

I think calling it view or span would be convey the wrong idea about what is going on.

Reply:
Sorry, by non-owning, I meant that it doesn't own the `SAFEARRAY` object (i.e. not responsible for calling `SafeArrayDestroy`). Alternative naming suggestions would be to include `ref`, `reference`, or `lock` into the data type name. `safearraydata` doesn't quite convey its ownership or responsibility.